src/workerd/jsg/setup.c++

```diff
@@ -152,145 +154,76 @@ void IsolateBase::clearDestructionQueue() {
-  // we have to make sure a TracedReference exists to tell V8 "this can't be collected without
-  // tracing".
-
-  // Calculate previous trace ID. See TracePrologue() wrap-around logic.
-  uint prevTraceId = traceId == 1 ? uint(kj::maxValue) : traceId - 1;
-
-  if (handle.lastMarked == prevTraceId) {
-    // Handle should still be valid.
-  } else {
-    // Handle was probably collected, or hadn't been created yet. (Re)create it.
-    v8::HandleScope scope(isolate);
-    kj::ctor(handle.tracedRef, isolate, handle.Get(isolate));
-  }
-
-  handle.lastMarked = traceId;
-
-  if (inAdvanceTracing) {
-    // We actually are in the middle of a trace, so mark the reference for V8.
-    RegisterEmbedderReference(handle.tracedRef);
-  } else if (inTrace) {
-    // A trace is happening but V8 hasn't called AdvanceTracing(), so we can't actually call back
-    // to RegisterEmbedderReference() right now as V8 will not like it. But we can't save a pointer
-    // to the handle itself to mark later becasue, who knows, maybe it'll be dropped on the C++
-    // side in the meantime. So we make a copy TracedReference which we can mark during the next
-    // AdvanceTracing(). Awkwardly, this means we don't actually mark `handle.tracedRef`, so it
-    // can still be collected, but at least the object it points to cannot be.
-    //
-    // Note that this code path only ever executes if incremental tracing is enabled, which
-    // as of this writing is disabled by default.
-    v8::HandleScope scope(isolate);
-    auto local = handle.Get(isolate);
-    referencesToMarkLater.add(v8::TracedReference<v8::Data>(isolate, local));
-
-    static bool logOnce KJ_UNUSED = ([]() {
-      // I can't figure out how to get this code to actually run in tests, but it is apparently
-      // happening in production and causing problems. Try to gather some stack traces from
-      // production.
-      KJ_LOG(ERROR, "mark() called during trace outside of AdvanceTracing?", kj::getStackTrace());
-      return true;
-    })();
-  }
+HeapTracer& HeapTracer::getTracer(v8::Isolate* isolate) {
+  return IsolateBase::from(isolate).heapTracer;
 }
 
 void HeapTracer::clearWrappers() {
   while (!wrappers.empty()) {
-    wrappers.front().detachWrapper();
+    // Don't freelist the shim because we're shutting down anyway.
+    wrappers.front().detachWrapper(false);
   }
+  clearFreelistedShims();
 }
 
-void HeapTracer::RegisterV8References(
-    const std::vector<std::pair<void*, void*>>& internalFields) {
-  for (auto [dummy, obj]: internalFields) {
-    KJ_ASSERT(dummy == obj);
-    wrappersToTrace.add(reinterpret_cast<Wrappable*>(obj));
-  }
+bool HeapTracer::IsRoot(const v8::TracedReference<v8::Value>& handle) {
+  // We can safely discard our wrapper objects, if they are unmodified, because we can recreate
+  // them later. If we see any other type of handle, conservatively assume it must be traced (so
+  // must be considered a root for minor GC).
```

Comment:
Nit: it's worth adding here that v8 will determine if the object is modified before calling IsRoot.